src/target/riscv/riscv-013.c

```diff
@@ -1952,6 +1952,81 @@ static int wait_for_idle_if_needed(struct target *target)
 	return ERROR_OK;
 }
 
+static int reset_dm(struct target *target)
+{
+	/* TODO: This function returns an error when a DMI operation fails.
+	 * However, [3.14.2. Debug Module Control] states:
+	 * > 0 (inactive): ... Any accesses to the module may fail.
+	 *
+	 * Ignoring failures may introduce incompatibility with 0.13.
+	 * See https://github.com/riscv/riscv-debug-spec/issues/1021
+	 */
+	dm013_info_t *dm = get_dm(target);
+	assert(dm && "DM is expected to be already allocated.");
+	assert(!dm->was_reset && "DM is reset only once.");
+	/* `dmcontrol.hartsel` should be read first, in order not to
+	 * change it when requesting the reset, since changing it
+	 * without checking that `abstractcs.busy` is low is
+	 * prohibited.
+	 */
+	uint32_t dmcontrol;
+	int result = dm_read(target, &dmcontrol, DM_DMCONTROL);
+	if (result != ERROR_OK)
+		return result;
+
+	if (get_field32(dmcontrol, DM_DMCONTROL_DMACTIVE)) {
+		/* `dmcontrol.hartsel` is not changed. */
+		dmcontrol = (dmcontrol & DM_DMCONTROL_HARTSELLO) |
+			(dmcontrol & DM_DMCONTROL_HARTSELHI);
+		LOG_TARGET_DEBUG(target, "Initiating DM reset.");
+		result = dm_write(target, DM_DMCONTROL, dmcontrol);
+		if (result != ERROR_OK)
+			return result;
+
```

Comment:
As already noted in the related debug spec thread (https://github.com/riscv/riscv-debug-spec/issues/1021#issuecomment-2093183331), I tend to believe that the requirement to not change the hartsel bits was meant for standard operation of the DM, not for master-reset of the whole DM.

I'd recommend to try to not change the hartsel bits, but if that is not possible, just go ahead and reset the debug module.

The dmactive bit is a master-reset of the DM, that means it is a mechanism how to bring the DM to a known initial state. So it would be very strange to add constraints, saying that this DM reset cannot be used under certain circumstances, even though the strict interpretation of the spec is saying exactly that.

In pseudo-code, my proposal would be:

```
read dmcontrol --> dmcontrol_val;
if (not successful) {
    LOG_TARGET_WARNING(target, "Could not read dmcontrol register. Attempting debug module reset anyway.");
    dmcontrol_val = 0x0;
}

if (dmcontrol.dmactive) {
    dmcontrol_val &= ~dmactive;
    write dmcontrol <-- dmcontrol_val;
    wait for DM reset acknowledge;
}

write dmcontrol <-- 0x1;
wait for DM activation after reset;

LOG_TARGET_DEBUG(target, "DM successfully reset.");
```

Reply:
Another reason for doing the reset is that (to my knowledge) no user has so far objected against the unconditional reset of the DM that riscv-openocd performs.

Reply:
Please, see my [comment in the related riscv-spec discussion](https://github.com/riscv/riscv-debug-spec/issues/1021#issuecomment-2093259615)

Reply:
> In pseudo-code, my proposal would be:
...

Your proposal is great. However, I would like to implement it as a separate commit.

This commit intends to address a different issue that I have introduced in 8319eee9e1ffc601b94b4223958180b49f8b8188:

I've merged these two writes:

https://github.com/riscv-collab/riscv-openocd/commit/8319eee9e1ffc601b94b4223958180b49f8b8188#diff-701864e2be3210275085c84b8f6fa7f9eb6171a2a18ca71638be42a155c08c63L1978
https://github.com/riscv-collab/riscv-openocd/commit/8319eee9e1ffc601b94b4223958180b49f8b8188#diff-701864e2be3210275085c84b8f6fa7f9eb6171a2a18ca71638be42a155c08c63L1985-L1987

Into one:

https://github.com/riscv-collab/riscv-openocd/commit/8319eee9e1ffc601b94b4223958180b49f8b8188#diff-701864e2be3210275085c84b8f6fa7f9eb6171a2a18ca71638be42a155c08c63R1977-R1979

This is not allowed by the spec:
> 0 (inactive): The module’s state, including authentication mechanism,
takes its reset values (the dmactive bit is the only bit which can be written to something other than its reset value).

And addressing this issue is the intention behind this commit.

Reply:
> I would like to implement it as a separate commit.

Sure, feel free to merge this one as is, and further refinement can take place in a separate PR. 

Thank you.